### PR TITLE
Use pki_types to improve the interoperability with the rustls ecosystem 

### PR DESCRIPTION
**Pull request by Tudyx**
Use [`pki_types`] to improve the interoperability with the `rustls` ecosystem and make types pretty clear, as suggested by https://github.com/rustls/rcgen/pull/170#issuecomment-1849565550. 

To fully get rid of the `pem` dependency (by replacing it by `rustls_pemfile`), I think those further actions will be needed:
- add `CertificateSigningRequestDer` to  [`pki_types`]
- add a `Csr` variant to [rustls_pemfile::Item] which is a newtype of ` CertificateSigningRequestDer` 
- add some encoding capability to [rustls_pemfile](https://docs.rs/rustls-pemfile/latest/rustls_pemfile/index.html#), basically it would mean adding headers ("---- BEGIN ----") and handle well line ending.

Please let me know if I miss something and if you think this goes in the right direction.
I'm open to submit following PR for this.

[rustls_pemfile::Item]: https://docs.rs/rustls-pemfile/latest/rustls_pemfile/enum.Item.html
[`CryptoProvider`]: https://docs.rs/rustls/0.22.2/rustls/crypto/struct.CryptoProvider.html
[`pki_types`]: https://docs.rs/rustls-pki-types/latest/rustls_pki_types/index.html#

### CODE REVIEW

**Comment by djc** [comment]
Thanks for working on this! I would propose doing a smaller increment which preserves the use of the pem crate for parsing (but still uses the pki-types for DER-containing types in the public API).

What do you think?

**Comment by Tudyx** [comment]
Yes sure, let's do this step by step!
I've kept the dependency to the `pem` crate, and submit a PR to `pki_types`  to add the `CertificateSigningRequestDer` type (https://github.com/rustls/pki-types/pull/33)

**Comment by cpu** [comment]
> ci / Validate external types appearing in public API (pull_request) Failing after 1m

To fix this task you'll need to update the cargo-check-external-types metadata to add `rustls_pki_types` to the allow-list:
https://github.com/rustls/rcgen/blob/747b5d8e72d1caadedff415445487320d899570a/rcgen/Cargo.toml#L47-L51

The intent of `pki-types` is to be a stable API and so it's also ignored for this task in the other various rustls repos.

**Comment by Tudyx** [comment]
> To fix this task you'll need to update the cargo-check-external-types metadata to add rustls_pki_types to the allow-list:
I see, thank you for the tip!

File: rcgen/tests/botan.rs
Comment (cpu) [inline rcgen/tests/botan.rs:21]: If you decide it's worth having the explicit lifetimes here and L26 could use the same treatment.

File: rcgen/tests/webpki.rs
Comment (cpu) [inline rcgen/tests/webpki.rs:57]: ditto lifetime feedback here and L71, L73

**Review by cpu — APPROVED** [review]
Thanks for picking this up. It'll be nice to get this in with the other breaking changes in `main` before a release :+1: 

I flagged some nits but the overall meat of the work seems great as-is.

**Review by cpu — APPROVED** [review]

**Comment by cpu** [comment]
We're going to cut a pki-types release with the CSR type: https://github.com/rustls/pki-types/pull/34 If you want to take a Cargo patch and switch this branch over we can probably have that release published before this PR is done with review saving a follow-up for that.

File: rcgen/src/lib.rs
Comment (Tudyx) [reply rcgen/src/lib.rs]: Thank you, I also think it's better this way.
 
Maybe we could activate the [`elided-lifetimes-in-paths`](https://doc.rust-lang.org/rustc/lints/listing/allowed-by-default.html#elided-lifetimes-in-paths) `rustc`  lint in another PR to enforce the `rustls` convention also on this crate?

**Comment by Tudyx** [comment]
I've used a patch while waiting for the release :+1:  

I wonder if `CeritificateSigningRequest::from_der`  should take ownership or a reference to `CeritificateSigningRequestDer`

File: rcgen/src/key_pair.rs
Comment (djc) [inline rcgen/src/key_pair.rs:135]: Maybe the docstring for this function should document that this function can only consume PKCS#8 keys?

File: rcgen/src/key_pair.rs
Comment (djc) [inline rcgen/src/key_pair.rs:179]: IMO we should take `impl Into<PrivatePkcs8Key<'_>>` instead to make it clear at compile time we only support PKCS#8 keys.

File: rcgen/src/key_pair.rs
Comment (djc) [inline rcgen/src/key_pair.rs:365]: PKCS#8?

File: rcgen/src/key_pair.rs
Comment (djc) [inline rcgen/src/key_pair.rs:121]: Hmm, maybe using `impl Into<>` here also kind of defeats the point?

File: rcgen/src/key_pair.rs
Comment (djc) [inline rcgen/src/key_pair.rs:178]: And same here.

File: rcgen/src/key_pair.rs
Comment (djc) [inline rcgen/src/key_pair.rs:132]: For these comments, maybe we should mention the `PRIVATE KEY` string used for the PEM delimiters, similar to what we have in rustls-pemfile?

**Review by djc — APPROVED** [review]

**Comment by cpu** [comment]
@est31 Do you think you'll have bandwidth to review this soon? Should we continue to hold merging?

**Comment by est31** [comment]
It needs a rebase, right?

File: rcgen/src/key_pair.rs
Comment (est31) [inline rcgen/src/key_pair.rs:139]: An into would also work here, right?

File: rcgen/src/key_pair.rs
Comment (est31) [inline rcgen/src/key_pair.rs:121]: Could you also mention in the docs that one can pass a byte slice and use `into`, maybe with an example?

File: rcgen/src/csr.rs
Comment (est31) [inline rcgen/src/csr.rs:49]: Could you also mention in the docs that one can pass a byte slice and use `into` maybe with an example?

File: rcgen/src/key_pair.rs
Comment (est31) [inline rcgen/src/key_pair.rs:182]: Could you also mention in the docs that one can pass a byte slice and use `into` maybe with an example?

File: rcgen/src/key_pair.rs
Comment (est31) [inline rcgen/src/key_pair.rs:226]: Could you also mention in the docs that one can pass a byte slice and use `into` maybe with an example?

File: rcgen/src/key_pair.rs
Comment (est31) [inline rcgen/src/key_pair.rs:367]: Why did you remove the old slice conversion functionality? It is easier to go through slices.

File: rcgen/src/key_pair.rs
Comment (est31) [inline rcgen/src/key_pair.rs:381]: Why did you remove the old Vec conversion functionality? It is easier to go through Vecs.

File: rcgen/src/lib.rs
Comment (est31) [inline rcgen/src/lib.rs:686]: Could you also mention in the docs that one can pass a byte slice and use `into` maybe with an example?

File: rcgen/src/lib.rs
Comment (est31) [inline rcgen/src/lib.rs:1256]: Could you also mention in the docs how this can be converted into a vec or byte slice so that it can be saved into a file, ideally with an example that we test or rustdoc links so that we notice the breakage.

File: rcgen/src/lib.rs
Comment (est31) [inline rcgen/src/lib.rs:1697]: Could you also mention in the docs how this can be converted into a vec or byte slice so that it can be saved into a file, ideally with an example that we test or rustdoc links so that we notice the breakage.

File: rcgen/src/lib.rs
Comment (est31) [inline rcgen/src/lib.rs:1718]: Could you also mention in the docs how this can be converted into a vec or byte slice so that it can be saved into a file, ideally with an example that we test or rustdoc links so that we notice the breakage.

**Review by est31 — CHANGES_REQUESTED** [review]
I've always felt down when using these special types with crates like ring for example (the rust crypto ecosystem also uses them heavily), because you have to import an additional type to use the API, and maybe even an additional crate. Actually you don't but you need to first look up if there is a from impl, etc. You also need to figure out how to convert it into bytes again in the output. This makes interaction with such APIs way more expensive for developers.

At the least, we should make this interaction very clear in the docs, making sure that users know how to use rcgen without using the interop crate. I've pointed out all the locations in the comments.

It might indeed improve interopability with other libraries/crates that use these types, but I think that most users just want to convert from byte slices and back, because they get something from a network, or a file, or the other way round. So while improving the interopability with a minority of use cases, it hurts the interopability with the majority (or at least I believe). Vec and slices have conversions to the `Bytes` type for example, but I couldn't find anything about `Bytes` in the docs of `PrivatePkcs8KeyDer`. So you first have to convert the `Bytes` to a Vec or a slice and then to a `PrivatePkcs8KeyDer`.

The main advantage I see in DER specific types would be that it could take care of conversion to PEM, you'd just call a function `.pem()` on it. But when I look at the rustdocs of [PrivatePkcs8KeyDer](https://docs.rs/rustls-pki-types/latest/rustls_pki_types/struct.PrivatePkcs8KeyDer.html#) I don't see any explanation how to convert it to pem, nor is there an easy `pem()` that returns a string (or another wrapper).

Such types do have advantages as they carry more meaning than the byte slice type, and I'm definitely not advocating for adopting C where everything is a `void*` but one should be extremely careful to not stand in the way of users, and stuff like I/O is usually just byte slices.

File: rcgen/src/key_pair.rs
Comment (Tudyx) [reply rcgen/src/key_pair.rs:379]: Basically for the same reasons mentioned in https://github.com/rustls/rcgen/pull/170
To my knowledge, they are at least 3 standard ways to store a key. PKCS#8 (RFC 5958),  PKCS#1 (RFC 3447) and Sec1 (RFC 5915). All the 3 can be encoded as BER or DER, so they can be represented by a `Vec<u8>/&[u8]`. 
By using a specific type, we make clear to the user what format is expected, which it could be easily missed if it was only written in the doc.

Another solution is to specify the encoding in the function name and take a `Vec<u8>`, like for the `String::from_utf16be` or `String::from_utf8`. We would have the same clarity but without the interoperability with the rustls ecosystem. That's why I think the suggested solution is better.

**Comment by djc** [comment]
> I've always felt down when using these special types with crates like ring for example (the rust crypto ecosystem also uses them heavily), because you have to import an additional type to use the API, and maybe even an additional crate. Actually you don't but you need to first look up if there is a from impl, etc. You also need to figure out how to convert it into bytes again in the output. This makes interaction with such APIs way more expensive for developers.

I do sympathize with these concerns, however in this case:

* I think it substantially helps users of different experience levels to be clear about PEM vs DER. Although rcgen already does a great job of making this clear in the function names, I think the type-level distinctions can help.
* It is very easy to convert a `&[u8]` or `Vec<u8>` to one of these new types.

(I myself recently got confused and passed DER into an API that expected a `Vec<u8>` containing PEM. This was an SDK wrapping tonic, which does clearly mark the underlying fields as PEM.)

> The main advantage I see in DER specific types would be that it could take care of conversion to PEM, you'd just call a function `.pem()` on it. But when I look at the rustdocs of [PrivatePkcs8KeyDer](https://docs.rs/rustls-pki-types/latest/rustls_pki_types/struct.PrivatePkcs8KeyDer.html#) I don't see any explanation how to convert it to pem, nor is there an easy `pem()` that returns a string (or another wrapper).

Yes, there is no PEM encoding code in the rustls project. I think we might want to change that going forward, and it would definitely make the case for this stronger.

File: rcgen/src/key_pair.rs
Comment (est31) [reply rcgen/src/key_pair.rs:379]: Ideally for the `&[u8]/Vec<u8>` path, we'd be sniffing the format and then convert it from there.

>  We would have the same clarity but without the interoperability with the rustls ecosystem.

I'm not arguing against existence of `From<PrivatePkcs8KeyDer<'_>>` here, that one can be kept. It's not an either or. Types can have multiple `From<...>` impls (that's the difference between type arguments for traits and associated types for traits, one cannot implement `Iterator` on the same type with multiple `Item`s).

**Comment by est31** [comment]
> I myself recently got confused and passed DER into an API that expected a `Vec<u8>` containing PEM. This was an SDK wrapping tonic, which does clearly mark the underlying fields as PEM.

Fair point, and I personally also had some confusing experiences where one Rust crypto library provided a key in one format and the other library was reading the key in the other format. In these instances, the chosen format has always been underdocumented (it's already over a year ago and I don't remember the exact details any more).

I think we can eventually merge this PR, if it documents the new API well, and by that I mean at each location where the new types are being used.

You are right that it's very easy to convert these types to/from byte buffers but it needs to be documented how, one shouldn't need the docs of the interface crate for that.

File: rcgen/src/key_pair.rs
Comment (Tudyx) [reply rcgen/src/key_pair.rs]: In my opinion, we should limit the way a user can construct the type, to improve consistency and reduce question about which methods should be used. But I don't feel strongly about it, so I restored the `&[u8]/Vec<u8>` in the latest push

**Comment by Tudyx** [comment]
I've rebased on the latest main and add some comments as suggested in the review.

File: rcgen/src/certificate.rs
Comment (est31) [inline rcgen/src/certificate.rs:285]:
```suggestion
	/// you have already a byte slice, just calling `into()` and taking a reference
```

File: rcgen/src/csr.rs
Comment (est31) [inline rcgen/src/csr.rs:50]:
```suggestion
	/// you have already a byte slice, just calling `into()` and taking a reference
```

File: rcgen/src/key_pair.rs
Comment (est31) [inline rcgen/src/key_pair.rs:165]:
```suggestion
	/// you have already a byte slice, just calling `into()` and taking a reference
```

File: rcgen/src/key_pair.rs
Comment (est31) [inline rcgen/src/key_pair.rs:228]:
```suggestion
	/// you have already a byte slice, just calling `into()` and taking a reference
```

File: rcgen/src/key_pair.rs
Comment (est31) [inline rcgen/src/key_pair.rs:280]:
```suggestion
	/// you have already a byte slice, just calling `into()` and taking a reference
```

**Review by est31 — APPROVED** [review]
good except the nits

**Comment by djc** [comment]
Thanks for working on this, I'll just merge this now and might follow up with a PR with some nits.

**Comment by cpu** [comment]
Nice work @Tudyx :clap:

**Pull request merged**
